Add walkthrough. (#28)
neuroinformatics-unit/course-ephys-osss: book/pipeline-walkthrough.qmd

Here we will go through the key steps at a high level, and develop the theory while adding more advanced methods and
use cases in the following chapters.

In this walkthrough, we will use a short example dataset available in the `data/pipeline-walkthrough`
of the [course repository](https://github.com/neuroinformatics-unit/course-ephys-osss/tree/main/book/data/pipeline-walkthrough).
Alterantively, the data can be downloaded from [here]() ADD ZENODO.

This recording is only 2.5 seconds long, from 384 channels on a 4-shank Neuropixels 2 probe acquired
in SpikeGLX. We use two of the four shanks, with 192 recording channels on each.

We will load and preprocess the data, run spike sorting and compute quality metrics for assessing the sorting output.

Let's get started!
 
## Data Loading and Exploration

### Data Loading

First, we will load the data with SpikeInterface.

SpikeInterface can load data from many different of acquisition systems, using functions within its [extractors module](https://spikeinterface.readthedocs.io/en/stable/modules/extractors.html). As our data is from SpikeGLX, we will use the 

> []`read_spikeglx`](https://github.com/neuroinformatics-unit/course-ephys-osss/tree/main/book/data/pipeline-walkthrough) method:

Reviewer: ```suggestion
[`read_spikeglx`](https://github.com/neuroinformatics-unit/course-ephys-osss/tree/main/book/data/pipeline-walkthrough) method:
```

```{python}
import spikeinterface.extractors as si_extractors

# This should be the path to the folder containing
# the data. On Windows, you may need to prepend
# r to the string e.g. r"C:\my\path"

> path_to_data_folder = "data/pipeline-walkthrough"

Reviewer: If we use pathlib, we can bypass the Windows stuff.
```
from pathlib import Path
path_to_data_folder = Path("data/pipeline-walkthrough")
```

works on all OSs

Author: The only problem is that in Windows the general workflow is to copy and paste paths directly from the file explorer in which case they come automatically as backslashes 🤢 so it would have to be `Path(r"C:\my\path")` should we use pathlib anyway as good practice?

Reviewer: Oh right - you still need to raw string it. So pathlib doesn't actually gain us anything here. Ok, just leave it.

recording = si_extractors.read_spikeglx(
    path_to_data_folder,
    stream_id="imec0.ap"
)

print(recording)
```

We need to tell SpikeInterface where the data is on our machine, and what [data stream](concept-notes.qmd#spikeglx-streams)💡 we want to load. If you are unsure of the stream, 
run the function without the `stream_id` argument and SpikeInterface will prompt you with the available streams.

> Because we want to load the AP data stream for spike sorting, we pass `"imec0.ap"`.

Reviewer: ```suggestion
Because we want to load the AP ("Action Potential") data stream for spike sorting, we pass `"imec0.ap"`.
```

:::

### Visualising the probe

There are many brands of recording probes used in extracellular electrophysiology (e.g. 
[Neuropixels](https://www.neuropixels.org/), 
[Cambridge Neurotech](https://www.cambridgeneurotech.com/neural-probes) or
[NeuroNexus](https://www.neuronexus.com/products/electrode-arrays/)
) which makes loading probes into a common framework complex.

Under the hood, SpikeInterface uses the [ProbeInterface](https://probeinterface.readthedocs.io/en/main/) package to load many different types of probe. In our case, it automatically identifies and loads the Neuropixels 2 (NP2) probe.

It is important to check that the probe has loaded as expected, and that the geometry matches how it was set up in your acquisition software. We can plot the probe to confirm it loaded correctly:

```{python}
from probeinterface.plotting import plot_probe
import matplotlib.pyplot as plt

probe = recording.get_probe()

plot_probe(probe)
plt.show()

```

As expected, we have a 4-shank Neurpixels probe in which all the recording channels are located in the middle two shanks.

### Visualising the data

> Electrophysiological recordings are susceptible to contamination from multiple noise sources: electrical interference, mechanical noise caused by probe movement, and artefacts arising from hardware defects or poor contact.

Reviewer: ```suggestion
Electrophysiological recordings are susceptible to contamination from multiple noise sources: electrical interference, mechanical noise caused by probe movement, and artefacts arising from hardware defects or poor contacts.
```

Poor contact**s**?  You mean the actual device contacts? Or contact with something??

Author: yup I meant some issue with the device contacts, maybe "broken contacts"? I guess this is a hardware defect, so maybe "hardware defects such as broken contacts"?

Also as an [ISSUE] for glossary we should include terms for electrode / contact and maybe pick one to use across the book if possible

Author: made issue and changed to "Electrophysiological recordings are susceptible to contamination from multiple noise sources: electrical interference, mechanical noise caused by probe movement, and artefacts arising from hardware defects such as broken or noisy recording electrodes."/

As such, it is important to visualise the data, especially after preprocessing, to check the data quality. First, lets's plot one second of data from our raw recording:

```{python}
import spikeinterface.widgets as si_widgets

start_time = recording.get_times()[0]

si_widgets.plot_traces(
    recording,
    time_range=(start_time, start_time + 0.1),
    return_in_uV=True,
    order_channel_by_depth=True
)
plt.show()
```

> 

Reviewer: I think you need to explain what the plot is showing: what's on y, what's on x, what does red/blue mean?

Author: nice added "The sorting output is written to disk. The output contains all of the output files of the sorter, and its organisation is sorter-dependent. As we as spike times and unit assignments, it contains other outputs from the sorter e.g. representative waveforms for each units ('templates')."

Reviewer: Oop - the plot I was referring to is the heatmap of the trace plot. That's what I think you need to explain.

Author: sorry I seem to have copy and pasted the wrong lines 😅 the correct text is:

"This plot is a heatmap view of the data: time in seconds is on the x-axis, and each row of the y-axis is the data from a separate recording channel (ordered by depth). The colour of the pixel indicate the voltage on that recording channel at that timepoint."

`return_in_uV` to ensure the data is returned in microvolts, as by default 
most electrophysiological recordings are [stored as unitless `int16` values](concept-notes.qmd#data-storage-in-int16-and-scaling-to-uv)💡.

`order_channel_by_depth` is used as, depending on the probe, the default ordering of channels is not always by depth, meaning the order may be scrambled on the plot.

:::{.callout-tip}
### Tip
In general when working with raw data, you can use [`recording.get_channel_locations()`](https://spikeinterface.readthedocs.io/en/latest/api.html#spikeinterface.core.BaseRecording.get_channel_locations) to inspect the default channel ordering of the 
channels, and [`order_channels_by_depth()`](https://spikeinterface.readthedocs.io/en/stable/api.html#spikeinterface.core.order_channels_by_depth) function can be used to get the indicies for reordering.

:::

> As we can see - the raw data looks terrible! It is severely contaminated by electrical noise. We will need to preprocess it to tidy it up.

Reviewer: I have no intuition of raw data plots. Does it look terrible? Why?? Is it obvious that it's contaminated by noise?

Maybe a better angle would be: we are looking for individual spikes, and we can't see any in the raw data. This is because the signal is contaminated with electrical noise etc, which we'll need to tidy up.

Author: good point how is "Our signal of interest are action potentials recorded on the probe, however, we cannot see any in the raw data! It is severely contaminated by noise e.g. electrical noise. We will need to preprocess the recording to reduce the noise and isolate our signal of interest." (added)

)
ax = plt.gca()
ax.set_ylim(200, 300)
plt.show()
```

:::{.callout-note icon=false}
### Discuss
Look carefully at the plot above. Why do you think there are predictable gaps in the signal every two channels?
:::

::: {.callout-note collapse="true"}
### Click to reveal the answer

Our recording channels are located on two shanks, however we have plot all recording channels together. The channels on different shanks have the same y-position, so they are plot next to eachother, but different x-positions. 

The shanks are too far apart for the same action potential to be recorded across both shanks. This means that the AP recorded on one shanks channels is interleaved with channels from the other shank on which no action potential is recorded.

:::

We can solve this by spliting the recording by shank using the the recording's `groups` property. 

When SpikeInterface loads the data, it will automatically assign different shanks to different groups. We can split by these groups, which will return a dictionary of two recordings (one for each channel group):

```{python}

split_recording = cmr_recording.split_by("group")

print(
    split_recording
)
```

> ::: {.callout-note warning}

Reviewer: ```suggestion
::: {.callout-warning}
```

        return_in_uV=True,
        order_channel_by_depth=True,
        ax=ax[shank_idx]
    )
    ax[shank_idx].set_title(f"Shank: {shank_idx}")

plt.show()
```

```{python}

si_widgets.plot_traces(
    split_recording[0],
    time_range=(23.923, 23.926),
    return_in_uV=True,
    order_channel_by_depth=True,
    mode="map"
)
ax = plt.gca()
ax.set_ylim(200, 300)
plt.show()
```

Much better!

Now we have our preprocessed data, and have visually confirmed the data quality is good, we are ready to perform spike sorting.

:::{.callout-tip}
### Things to try

- How does the data look using `median` vs. `mean` as the `operator` argument in `common_average()`?
- What happens if you set the bandpass filter minimum cutoff to zero? What frequencies are including now? Or set the maximum cutoff to 15000?
- Set `return_in_uV=False`. Does the data look much different?
- Save the data to disk with the Recording objects `.save()` attribute
- Use `get_traces()` with and without `return_in_uV` as `True` and `False` to inspect the data type.

***stretch goal**

- Perform spike detection (e.g. absolute threshold method) on the preprocessed data obtained with `get_traces()`

:::

## Sorting

> Now we are ready to move onto spike sorting. The goal of spike sorting is to take our preprocessing recording with clear action potentials, and detect these action potentials and sort them as coming from different neurons, based on their waveform and positio on the probe.

Reviewer: ```suggestion
Now we are ready to move onto spike sorting. The goal of spike sorting is to take our preprocessing recording with clear action potentials, and detect these action potentials and sort them as coming from different neurons, based on their waveform and position on the probe.
```

> We can pass the recording to SpikeInterface's []`run_sorter()`](https://spikeinterface.readthedocs.io/en/stable/api.html#spikeinterface.sorters.run_sorter) function, which allows us to easily switch between different sorters. There are [many possible spike sorters](https://spikeinterface.readthedocs.io/en/stable/modules/sorters.html#supported-spike-sorters) to choose from, today we will use Kilosort 4.

Reviewer: ```suggestion
We can pass the recording to SpikeInterface's [`run_sorter()`](https://spikeinterface.readthedocs.io/en/stable/api.html#spikeinterface.sorters.run_sorter) function, which allows us to easily switch between different sorters. There are [many possible spike sorters](https://spikeinterface.readthedocs.io/en/stable/modules/sorters.html#supported-spike-sorters) to choose from, today we will use Kilosort 4.
```

> Kilosort 4 by default will perform similar preprocessing steps to which we have already applied (it does highpass filtering and common average referencing).

Reviewer: ```suggestion
By default, [Kilosort 4](https://github.com/MouseLand/Kilosort) will perform similar preprocessing steps to which we have already applied (it does highpass filtering and common average referencing).
```

> As such, we will skip these steps, as we have already run these steps in SpikeInterface (we cannot turn filtering off, so just set it to a very low cutoff).

Reviewer: I think you can turn off filtering, by doing

```
skip_kilosort_preprocessing = True
```

to `run_sorter`

Author: I think this will also turn off CAR and whitening (as sets Wrot to identitiy [here](https://github.com/SpikeInterface/spikeinterface/blob/dbe2e7539ab63ebc0c3ddc856bfbc93f10e067aa/src/spikeinterface/sorters/external/kilosort4.py#L348))

Author: actually the CAR is still based on `do_CAR` which is maybe confusing? you might expect it to not CAR whatever `do_CAR` is if `skip_kilosort_preprocessing =True`

Reviewer: Oh, didn't realised it turned off whitening! Ok, your notes are correct.

It is generally easiest to sort each shank separately, and in this tutorial we will now sort the first shank of the recording:

```{python}
from pathlib import Path
import spikeinterface.sorters as si_sorters

sorter_folder = Path("kilosort4_output")
sorter_output_folder = sorter_folder / "sorter_output"

first_shank_recording = split_recording[0]

sorting = si_sorters.run_sorter(
    "kilosort4",
    first_shank_recording,
    folder=sorter_folder,
    remove_existing_folder=True,
    do_CAR=False,

>     highpass_cutoff=0.1,

Reviewer: ```suggestion
    skip_kilosort_preprocessing = True
```

)

print(sorting)
```

> The sorting output is written to disk. The output contains all of the output files of the sorter, and its organisation is sorter-dependent. As we as spike times and unit assignments, it contains other outputs from the sorter e.g. representative waveforms for each units ('templates').

Reviewer: ```suggestion
The sorting output is written to disk. The output contains all of the output files of the sorter, and its organisation is sorter-dependent. As well as spike times and unit assignments, it contains other outputs from the sorter e.g. representative waveforms for each units ('templates').
```

In SpikeInterface, we can access the spike times and unit assignments in a
sorter-independent manner, using the [`sorting`](https://spikeinterface.readthedocs.io/en/latest/api.html#spikeinterface.core.BaseSorting) object. For example, to
get the spike times and unit assignments:

> 

Reviewer: Think it'd be good to print the sorting here-ish

```{python}

spikes = sorting.to_spike_vector()

spike_times = spikes["sample_index"] / sorting.get_sampling_frequency()

# show the first 5 spikes
print(
    f"`spikes` is a structured number array with fields`: {spikes.dtype.names}\n"
)
print(
    f"First 5 spike times (s): {spike_times[:5]}\n"
)
print(
    f"First 5 unit assignments: {spikes['unit_index'][:5]}"
)

```

:::{.callout-tip}
### Things to try
 - Get a list of all non-empty unit ids
 - Remove a unit - what happens to the spikes assigned to this unit?
 - Get the spike times for a specific unit
:::

## Postprocessing

> 

Reviewer: I think on the actual day, we should compute a load of stuff and switch now to SpikeInterface-GUI. I have a nice demo with raw data which can demonstrate everything detailed here. I think the interactive nature will get the point across better. Let's take a look IRL?

But I would keep these notes as a reference for people to read, and see how to do more customised computations etc.

EDIT: Actually, I'm not even sure if we need these notes here: some of them are very detailed. Maybe we can port them over to Alessio's section?

Author: Yes I agree that would work nicely on the day, maybe we can make a standalone slide or a dropdown here with the code for the extension computation.

Here we can have sections for selecting spikes, waveforms, templates and drop amplitudes, and for each section just have 1-2 sentences explaining what it is, and a picture for waveforms and templates?

Alternatively, we could have a very short code block here that just computes a load of extensions and introduces the analyzer. In text we could have an brief explanation that we are selecting spikes, and computing things like waveforms, templates, amplitudes. Then we can link to glossary / concept notes for these from here, and also from everywhere else relevant?

To determine whether units are good or not, we must look at the individual spikes that belong to a unit, and compute metrics. Cut then out, waveforms. In this section we will go over some key concepts at a very high level - we will develop this much further in the Postprocessing chapter.

### Creating a `SortingAnalzyer`

The `SortingAnalzyer` is the key object for postprocessing in SpikeInterface. We can create the `SortingAnalzyer` by passing the sorting output and preprocessed recording:

```{python}
from spikeinterface import create_sorting_analyzer

analyzer = create_sorting_analyzer(
    sorting=sorting,
    recording=first_shank_recording
)
```

Using this sorting analyzer interface, we will compute all of the
postprocessing results needed to assess the quality of our sorting.

:::{warning}

Some sorters will output their own waveforms, templates, amplitudes ETC (for example, Kilosort has a `templates.npy` file containnig the templates it generated during sorting).

These are not loaded by SpikeInterface. Instead, the sorting analzyer will always recompute these features directly from the passed `recording`.

:::

### Selecting spikes

> In our sorting, we can have many many detected spikes. This may be in the order of millios of spikes for an hour long recording.

Reviewer: ```suggestion
In our sorting, we can have many many detected spikes. This may be in the order of millions of spikes for an hour long recording. There are very long recordings from the Sainsbury Wellcome Centre which have billions of spikes!
```

Author: good point! slightly reworded in an effort to keep the book location-agnostic? "In our sorting, we can have many many detected spikes. This may be in the order of millions of spikes for an hour long recording. When performing experiments with very long recordings (e.g. days or weeks) billions of spikes may be detected!" (added)

Reviewer: Oh, I was assuming the opposite - we should get some SWC bragging/branding in the book! Fair enough to keep it agnostic.

In general, we will compute summary statistics over metrics computed from individual spikes. Therefore, including millions of spikes is computationally intensive while providing marginal gains. Instead, we can subselect spikes to include in postprocessing computations (typically, 3-500 spikes per unit is used a default).

```{python}
analyzer.compute(
    "random_spikes",
    method="uniform",
    max_spikes_per_unit=500,
)
```

To access the data assicated with a computed extension, we can use the `get_extension` function as below. We can see the indices (and compute the 
spike times) for the selected spikes:

```{python}
spike_indices = analyzer.get_extension("random_spikes").data["random_spikes_indices"]
spike_times = spike_indices / recording.get_sampling_frequency()

print(spike_times)
```

### Waveforms

The action potential waveforms are important for assessing unit quality. For example, we expect that the waveforms of spikes which belong to a single unit are all similar.

> To compute waveforms in SpikeInterface, we can use the `waveforms` extension. This will 'chop out' the spike waveform from the data, with a time window specified by the `ms_before` and `ms_after` (the spike peak) arguments. For example in the schematic below, we could cut out the blue spike in a window defined by the red blox:

Reviewer: ```suggestion
To compute waveforms in SpikeInterface, we can use the `waveforms` extension. This will 'chop out' the spike waveform from the data, with a time window specified by the `ms_before` and `ms_after` (the spike peak) arguments. For example in the schematic below, we could cut out the blue spike in a window defined by the red box:
```

Author: blox :( you will be glad to know I have installed a spell checker on vscode

![Example raw trace segment with an artifact highlighted in red.](images/intro-waveforms.png)

```{python}
analyzer.compute("waveforms", ms_before=1.0, ms_after=2.0)
```

> The waveforms array is three dimensional `(num waveforms, num samples, num channels)

Reviewer: ```suggestion
The waveforms array is three dimensional `(num waveforms, num samples, num channels)`
```

```{python}

> waveforms = analyzer.get_extension("waveforms").data["waveforms"]

Reviewer: ```suggestion
waveforms = analyzer.get_extension("waveforms").get_data()
```

We've been vaguely trying to encourage use of `get_data` cause it works for all extensions. `data` contains lots of junk: computation time etc.

Author: nice will use that in future

print(waveforms.shape)

# This is the waveform of the 19th spike (chosen as it looks nice)

> plt.imshow(waveforms[18, :, :].T, aspect="auto")

Reviewer: Maybe use a colour scheme to match the red and blue from earlier?

plt.show()
```

### Templates

> The template is a representative wavecform from a putative neuron. Often, it is computed as the average waveform from all of the spikes that were assigned to a unit.

Reviewer: ```suggestion
The template is a representative waveform from a putative neuron. Often, it is computed as the average waveform from all of the spikes that were assigned to a unit.
```

We can compute the extension and access the templates with:

```{python}
analyzer.compute("templates")

> templates = analyzer.get_extension("templates").data["average"]

Reviewer: ```suggestion
templates = analyzer.get_extension("templates").get_data()
```

print(
    templates.shape
)
```

SpikeInterface also includes many useful post-processing plots in the widget smodule. For example  we can plot a unit template as below:

```{python}

# Plot the template waveform of the first unit 
si_widgets.plot_unit_templates(analyzer, unit_ids=[0])
plt.show()

```

### Amplitudes

> Finally, spike amplitudes are a useful measure for assessing unit quality. We expect that the amplitude of a spike is relatively stable througghout a recording. Changes in spike amplitudes may indicate problems with hte neuron health, drift, or other issues. 

Reviewer: ```suggestion
Finally, spike amplitudes are a useful measure for assessing unit quality. We expect that the amplitude of a spike is relatively stable throughout a recording. Changes in spike amplitudes may indicate problems with the neuron health, drift, or other issues. 
```

To compute the spike amplitudes, we can run:

```{python}

analyzer.compute("spike_amplitudes")

> amplitudes = analyzer.get_extension("spike_amplitudes").data["amplitudes"]

Reviewer: ```suggestion
amplitudes = analyzer.get_extension("spike_amplitudes").get_data()
```

print(amplitudes)
```

This gives a list of amplitudes, one for every spike. We will explore how visualisng spike amplitude across time is a useful measure for quality asessment later.

## Manual Curation in spikeinterface-gui

From this, we can compute quality metrics used to filter the units e.g. bombcell. We cover this more in the postprocessing chapter.

For now, see SpikeInterface GUI video?

> https://www.youtube.com/watch?v=OGlQtxzip-Q

Reviewer: Edit this later.
